### PR TITLE
basic certificate handling for quic

### PR DESCRIPTION
**Pull request by wfurt**
With https://github.com/microsoft/msquic/pull/1411 support for STUB Tls was removed and we only have real option left. 

This change adds server support for normal X509Certificate and plus usual verification callback so the fate of the handshake can be decided from .NET. 

Right now, this is supported only ion Windows, Linux & macOS coming (hopefully) soon.

This is also missing support for custom certificate chains, SslStreamCertificateContext, and client certificates. 

cc: @nibanks

### CODE REVIEW

**Comment by wfurt** [comment]
Basic certificates should now work on Linux & macOS well. 
Some networking tests are failing like this on macOS
```
System.Net.Quic.Tests.QuicConnectionTests_MsQuicProvider.TestConnect [FAIL]
        Assert.Equal() Failure
        Expected: 127.0.0.1:63445
        Actual:   [::]:63445
```
I will investigate in separate effort.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs [hunk 1]
```diff
@@ -272,6 +275,17 @@ internal struct CredentialConfigCertificateCertificateFileProtected
             internal string PrivateKeyPassword;
         }
 
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct CredentialConfigCertificatePkcs12
+        {
+            internal IntPtr Asn1Blob;
+
+            internal int Asn1BlobLength;
```
Comment (nibanks) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs:283]: shouldn't this be a `uint`?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicConfigurationHandle.cs [hunk 2]
```diff
@@ -59,6 +60,18 @@ private static unsafe SafeMsQuicConfigurationHandle Create(QuicOptions options,
...
+            if ((flags & QUIC_CREDENTIAL_FLAGS.CLIENT) == 0)
+            {
+                if (certificate == null)
+                {
+                    throw new Exception("Server must provide certificate");
+                }
+            }
+            else
+            {
+                flags |= QUIC_CREDENTIAL_FLAGS.INDICATE_CERTIFICATE_RECEIVED | QUIC_CREDENTIAL_FLAGS.NO_CERTIFICATE_VALIDATION;
```
Comment (nibanks) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicConfigurationHandle.cs:72]: You never want to let the TLS stack handle default validation?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs [hunk 3]
```diff
@@ -181,6 +199,74 @@ private static uint HandleEventStreamsAvailable(State state, ref ConnectionEvent
...
+            }
+
+            if (connectionEvent.Data.PeerCertificate.PlatformCertificateHandle != IntPtr.Zero)
+            {
+                certificate = new X509Certificate2(connectionEvent.Data.PeerCertificate.PlatformCertificateHandle);
+            }
+
+            try
+            {
+                if (certificate == null)
```
Comment (nibanks) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:227]: I don't expect this to ever happen. We shouldn't be calling you with "Peer Certificate Received" if we didn't have a certificate. Personally, I'd say you could just assert here or something.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs (hunk 1 above)
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs:283]: That is generally the default for .NET -> use int unless there is reason no to. I personally don't care either way.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicConfigurationHandle.cs (hunk 2 above)
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicConfigurationHandle.cs:72]: That how SslStream works now. Long term, we may check if custom callback is present or now. But that is still problematic (at least on Linux) as X509Chain would consider user stores while OS does not have any clue about them. So we would end up in situation when SslStream works and Quick does not or vice versa.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs (hunk 3 above)
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:227]: The null makes perfect sense for server side where client's certificate is optional. I just need to complete that part but the validation callback will be shared.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs (hunk 1 above)
Comment (danmoseley) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs:283]: We mainly use uint when necessary to match native headers.

**Comment by nibanks** [comment]
This is supposed to match our maybe header, which uses uint32_t.

Get Outlook for Android<https://aka.ms/AAb9ysg>

________________________________

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicConfigurationHandle.cs [hunk 4]
```diff
@@ -103,27 +116,42 @@ private static unsafe SafeMsQuicConfigurationHandle Create(QuicOptions options,
...
+                        config.Certificate = certificate.Handle;
+                        status = MsQuicApi.Api.ConfigurationLoadCredentialDelegate(configurationHandle, ref config);
+                    }
+                    else
+                    {
+                        CredentialConfigCertificatePkcs12 pkcs12Config;
+                        byte[] asn1 = certificate.Export(X509ContentType.Pkcs12);
+                        unsafe
+                        {
+                            fixed (void* ptr = asn1)
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicConfigurationHandle.cs:136]: I think this is the only occurrence of `fixed` in here. I'm not an interop expert, but should we rather use `GCHandle` as we do in other places? Is there any difference in those two? If not should we rather stick to the same approach throughout?
cc: @scalablecory

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs [hunk 5]
```diff
@@ -181,6 +199,74 @@ private static uint HandleEventStreamsAvailable(State state, ref ConnectionEvent
...
 
+        private static uint HandleEventPeerCertificateReceived(State state, ref ConnectionEvent connectionEvent)
+        {
+            SslPolicyErrors sslPolicyErrors  = SslPolicyErrors.None;
+            X509Chain? chain = null;
+            X509Certificate2? certificate = null;
+
+            if (!OperatingSystem.IsWindows())
+            {
+                // TBD fix validation with OpenSSL
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:210]:
```suggestion
                // TODO: fix validation with OpenSSL
```

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs [hunk 6]
```diff
@@ -30,6 +35,10 @@ internal sealed class MsQuicConnection : QuicConnectionProvider
         private IPEndPoint? _localEndPoint;
         private readonly EndPoint _remoteEndPoint;
         private SslApplicationProtocol _negotiatedAlpnProtocol;
+        private bool _isServer;
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:38]: Could any of these fields be readonly?
Do we need both: `_isServer` and `_remoteCertificateRequired`? From the ctors it looks like they're just opposites.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs [hunk 7]
```diff
@@ -181,6 +199,74 @@ private static uint HandleEventStreamsAvailable(State state, ref ConnectionEvent
...
+                    sslPolicyErrors |= SslPolicyErrors.RemoteCertificateNotAvailable;
+                }
+                else
+                {
+                    chain = new X509Chain();
+                    chain.ChainPolicy.RevocationMode = connection._revocationMode;
+                    chain.ChainPolicy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
+                    chain.ChainPolicy.ApplicationPolicy.Add(connection._isServer ? s_clientAuthOid : s_serverAuthOid);
+
+                    if (!chain.Build(certificate)) {
```
Comment (ManickaP) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:239]:
```suggestion
                    if (!chain.Build(certificate))
                    {
```

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicEnums.cs [hunk 8]
```diff
@@ -19,6 +19,7 @@ internal enum QUIC_CREDENTIAL_TYPE : uint
         CONTEXT,
         FILE,
         FILE_PROTECTED,
+        PKCS12,
         STUB_NULL = 0xF0000000, // Pass as server cert to stubtls implementation.
```
[the remark this answers is not in the document]
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicEnums.cs:23]: we may. I'm wondering if this is useful for anybody who is on older version of msquic. Adding certificate support is primary goal. Removal of stub not so much.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs (hunk 6 above)
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:38]: for client, _remoteCertificateRequired is always true. For server this is optional depending on setting. That part is not wired completely as client certs are not fully functional.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs (hunk 1 above)
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs:283]: array.Length is `int` so as pretty much all `Length` properties in c#. We will have to cast and the outcome would be same. 
This is pretty common that underlying `C` uint is exposed as `int` to make c# code more natural. For example 

https://github.com/dotnet/runtime/blob/0423ac97a6dcb54a9f096f848cd20346217f0c6a/src/libraries/Common/src/Interop/Unix/System.Security.Cryptography.Native/Interop.Ssl.cs#L54-L55

and 

https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/Native/Unix/System.Security.Cryptography.Native/pal_ssl.h#L366

cc: @stephentoub for any thoughts.  I'm fine if we embark on the casting route as this is all internal.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs [hunk 9]
```diff
@@ -434,7 +456,10 @@ internal struct ConnectionEventDataUnion
             [FieldOffset(0)]
             internal ConnectionEventDataStreamsAvailable StreamsAvailable;
 
-            // TODO: missing IDEAL_PROCESSOR_CHANGED, ..., PEER_CERTIFICATE_RECEIVED (7 total)
+            [FieldOffset(0)]
+            internal ConnectionEventPeerCertificateReceived PeerCertificate;
```
[the remark this answers is not in the document]
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs]: updated.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicConfigurationHandle.cs [hunk 10]
```diff
@@ -103,27 +116,42 @@ private static unsafe SafeMsQuicConfigurationHandle Create(QuicOptions options,
...
+                    {
+                        config.Type = QUIC_CREDENTIAL_TYPE.CONTEXT;
+                        config.Certificate = certificate.Handle;
+                        status = MsQuicApi.Api.ConfigurationLoadCredentialDelegate(configurationHandle, ref config);
+                    }
+                    else
+                    {
+                        CredentialConfigCertificatePkcs12 pkcs12Config;
+                        byte[] asn1 = certificate.Export(X509ContentType.Pkcs12);
+                        unsafe
```
[the remark this answers is not in the document]
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicConfigurationHandle.cs]: no, we don't

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicConfigurationHandle.cs (hunk 4 above)
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/SafeMsQuicConfigurationHandle.cs]: You need to allocate GCHandle to preserve object for longer time. This basically says "don't move asn1 while I'm in this block" and it is much simpler. (and does not work for the other cases.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicEnums.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (scalablecory) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicEnums.cs:23]: We should just remove it if msquic doesn't support it. It's a shame though; there is no other way to test on non-beta Windows :(

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs (hunk 1 above)
Comment (scalablecory) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs:283]: I'm okay with us using `int` for buffer lengths. I don't know why `uint` is being used so excessively in the code right now, it doesn't really add value. Whatever we do we should be consistent.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs [hunk 11]
```diff
@@ -407,6 +421,14 @@ internal struct ConnectionEventDataStreamsAvailable
             internal ushort UniDirectionalCount;
         }
 
+        [StructLayout(LayoutKind.Sequential)]
+        internal struct ConnectionEventPeerCertificateReceived
+        {
+            internal IntPtr PlatformCertificateHandle;
+            internal uint DeferredErrorFlags;
+            internal int DeferredStatus;
```
[the remark this answers is not in the document]
Comment (scalablecory) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs:429]: Yeah this is where I'd like us to keep using `uint` (assuming this is what MsQuic uses) -- it's not a buffer length, and it's annoying to see `-1234` in a debugger but `0xF.....` in a header.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs:283]: One of our goals with our interop files is that they should be as close to the native API signature as possible,  e.g. as if we were able to just include the .h file and target its signature directly.  So we've used uint (or at least intended to) where the underlying signature is for an unsigned int,  even if that complicates the interop consumption slightly.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs [hunk 12]
```diff
@@ -181,6 +199,75 @@ private static uint HandleEventStreamsAvailable(State state, ref ConnectionEvent
...
 
+        private static uint HandleEventPeerCertificateReceived(State state, ref ConnectionEvent connectionEvent)
+        {
+            SslPolicyErrors sslPolicyErrors  = SslPolicyErrors.None;
+            X509Chain? chain = null;
+            X509Certificate2? certificate = null;
+
+            if (!OperatingSystem.IsWindows())
+            {
+                // TODO fix validation with OpenSSL
```
Comment (scalablecory) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:210]: What is remaining to do here? Is there an issue?

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs [hunk 13]
```diff
@@ -181,6 +199,75 @@ private static uint HandleEventStreamsAvailable(State state, ref ConnectionEvent
...
+                if (!connection._remoteCertificateRequired)
+                {
+                    sslPolicyErrors &= ~SslPolicyErrors.RemoteCertificateNotAvailable;
+                }
+
+                if (connection._remoteCertificateValidationCallback != null)
+                {
+                    bool success = connection._remoteCertificateValidationCallback(connection, certificate, chain, sslPolicyErrors);
+                    if (!success && NetEventSource.Log.IsEnabled())
+                        NetEventSource.Error(state.Connection, $"Remote certificate rejected by verification callback");
```
Comment (scalablecory) [inline src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:254]:
```suggestion
                        NetEventSource.Error(state.Connection, "Remote certificate rejected by verification callback");
```

**Review by scalablecory — APPROVED** [review]

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicEnums.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicEnums.cs:23]: ok, I'll remove it.
On a side note, did you have success decrypting the traffic with Wireshark @nibanks? 
Being able to peek inside of the streams has benefits.

[the remark this answers is not in the document]
Comment (nibanks) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicEnums.cs:23]: We have `QUIC_PARAM_CONN_TLS_SECRETS` in https://github.com/microsoft/msquic/blob/main/src/inc/msquicp.h#L105. This allows the app to get the TLS secrets out of the QUIC stack (assuming you build with this feature enabled). You can then give those to wireshark to decode traces. The interop runner setup does this. Feel free to IM me for more details.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs (hunk 12 above)
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/MsQuicConnection.cs:210]: The issue is that we don't have good way how to get certificate chain right now with OpenSSL. 
There is not specific issue AFAIK. I mentally do this as contributions to  #49574.
I will follow-up if/when https://github.com/microsoft/msquic/pull/1450 (+implementation) is ready.

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicEnums.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicEnums.cs]: removed. I will update README in separate change (and also add notes about macOS)

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs (hunk 1 above)
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs]: updated to `uint`

File: src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs (hunk 11 above)
[the remark this answers is not in the document]
Comment (wfurt) [reply src/libraries/System.Net.Quic/src/System/Net/Quic/Implementations/MsQuic/Interop/MsQuicNativeMethods.cs]: updated to `uint`

**Comment by wfurt** [comment]
all the concerns should be addressed. Following outstanding issues will be follow-up changes:
- certificate chain handling
- client certificates
- certificate validation with OpenSSL

**Review by nibanks — APPROVED** [review]

**Comment by JamesNK** [comment]
I created a number of TLS issues for QUIC. It would be good to know which ones your PR makes progress on.

For https://github.com/dotnet/runtime/issues/49574, it looks like this checks off RemoteCertificateValidationCallback and CertificateRevocationCheckMode.

Does this PR start to unblock https://github.com/dotnet/runtime/issues/50156?

**Comment by wfurt** [comment]
With this, self-signed (and other certificates) will work on Windows @JamesNK. Linux/macOS is getting closer. 
I'm planning to go through the open issues once we have platform parity.

**Pull request merged**